src/sentry/search/django/backend.py

```diff
@@ -272,14 +272,28 @@ def query(self, project, tags=None, environment=None, sort_by='date', limit=100,
         retention = quotas.get_event_retention(organization=project.organization)
         if retention:
             retention_window_start = timezone.now() - timedelta(days=retention)
-            # TODO: This could be optimized when building querysets to identify
-            # criteria that are logically impossible (e.g. if the upper bound
-            # for last seen is before the retention window starts, no results
-            # exist.)
-            group_queryset = group_queryset.filter(last_seen__gte=retention_window_start)
         else:
             retention_window_start = None
 
+        # This is a punt because the SnubaSearchBackend (a subclass) shares so much that it
+        # seemed better to handle all the shared initialization and then handoff to the
+        # actual backend.
+        return self._query(project, retention_window_start, group_queryset, tags,
+                           environment, sort_by, limit, cursor, count_hits,
+                           paginator_options, **parameters)
+
+    def _query(self, project, retention_window_start, group_queryset, tags, environment,
+               sort_by, limit, cursor, count_hits, paginator_options, **parameters):
+
+        from sentry.models import (Group, Environment, Event, GroupEnvironment, Release)
+
+        # TODO: This could be optimized when building querysets to identify
+        # criteria that are logically impossible (e.g. if the upper bound
+        # for last seen is before the retention window starts, no results
+        # exist.)
+        if retention_window_start:
+            group_queryset = group_queryset.filter(last_seen__gte=retention_window_start)
```

Comment:
Shouldn't this be applied to all `Group` queries, not just Django ones?

Reply:
Retention window is respected as part of Snuba's required `start` and `end` parameters. It ends up being equivalent, we just don't lean on the denormalized `Group` field. Though I did some cleanup/fix it after you mentioned that:

https://github.com/getsentry/sentry/pull/7925/files#diff-831b13b0865daf8518e97b6b74ae6d53R158

Reply:
I was more thinking along the lines that if we are going to keep `Group.last_seen` around it might make sense to push this filter down anyway just to reduce the number of hashes that need to get sent to the search backend.